src/cli/cli.cpp

```diff
@@ -0,0 +1,123 @@
+#include "cli.hpp"
+#include "states.hpp"
+#include "Particle.h"
+#include "conio.hpp"
+#include "menuItems/systemCommands.hpp"
+
+#include <fstream>
+
+const CLI_menu_t CLI_menu[] =
+{
+    {"menu", "display Menu", &CMD_displayMenu},
+    {"disconnect", "disconnect particle", &CLI_disconnect},
+    {"connect", "connect particle", &CLI_connect},
+    {"showerrors", "show flog errors", &CLI_displayFLOG},
+    {NULL,NULL,NULL}
+};
+
+static STATES_e CLI_nextState;
+
+void CLI::init(void) 
+{
+    CLI_nextState = STATE_CLI;
+
+    // While there is an avaliable character typed, get it
+    while(kbhit())
+    {
+        getch();
+    }
+}
+
+STATES_e CLI::run(void)
+{
+    CLI_menu_t *cmd;
+    char* userInput = new char[100];
+    userInput[0] = 0;
+
+    SF_OSAL_printf("\n>");
+
+    while (1) 
+    {
+        //TODO: check for input timeout
+        char ch;
+        int i = 0;
+
+
+        bool finishedTyping = false;
+        // Loop through user input until they finish typing their command
+        while (!finishedTyping) 
+        {
+            if(kbhit()) 
+            {
+                ch = getch();
+
+                if(i > 99) {
+                    // too long of a command
+                    SF_OSAL_printf("\n\nCommand too long");
+
+                    break;
+                }
+
+                switch(ch) 
+                {
+                    case '\b':
+                        i--;
+                        SF_OSAL_printf("\b \b");
+                        userInput[i] = '\0';
+                        break;
+                    case '\r':
+                    case '\n':
+                        userInput[i++] = 0;
+                        finishedTyping = true;
+                        break;
+                    default:
+                        putch(ch);
+                        userInput[i++] = ch;
+                        break;
+                }
+            } else{
+                continue;
+            }
+        }
+
+        if(strlen(userInput) != 0) 
+        {
+            cmd = CLI_findCommand(userInput);
+            if(!cmd) 
+            {
+                putch(userInput[0]);
+                SF_OSAL_printf("Unknown command\n\n");
+                SF_OSAL_printf(">");
+            } else {
+                cmd->fn();
+                SF_OSAL_printf("\n>");
+            }
+        }
+        
+        userInput = new char[100];
```

Comment:
Memory leak - the previously allocated `char` array has not been freed and is now lost!